### DISCUSSION TITLE
support SFZ 'script' opcode

### DISCUSSION THREAD

**Discussion by grammoboy2** [sfztools/sfizz#1215] [Requests]
Support sfz opcode 'script' for NKSP scripts etc.

http://doc.linuxsampler.org/Instrument_Scripts/NKSP_Language/

**Comment by jpcima** [comment]
This issue has been discussed several times, but no progress currently.
There is an idea of how to implement it and make it efficient, but perhaps first it's best to define the scope first.
What are requirements and typical use cases of a script language?

**Comment by kinwie** [comment]
Iirc, first use of kontakt scripts are for the midi preprocessor, e.g. arpeggiator, harmonizer, drums step sequencer, etc. And also for the "auto" things, like auto legato, auto portamento, etc, as well as other functions like RR simulation and placing common used CC at gui. Then, a lot of things can be created with scripts.
If sfz can use more advanced scripts, then it will be fantastic. Though, I haven't had any pictures of how the mechanism will works with current state of sfz.

**Comment by kinwie** [comment]
My short understanding of scripts usage is, the scripts can be pre-made by anybody and act like a script 'preset. Then user can put the script into sfz or include it to apply the function.
So yeah, a choosen language is needed to build the script and get recognized by sfz engine.
I would prefer a more easier programming language follow 'sfz principal', the human readable lines :)

**Comment by kinwie** [comment]
An example use case that will be good for sfz engine as a drum sampler is the note-remapping, so a script that allow freely remapping the drum notes via GUI

**Comment by grammoboy2** [comment]
From
https://nilsliberg.se/ksp/scripts/tutorial/

"What are scripts?

Scripts can be seen as small plugins. They are quite similar to the MIDI effect plugins you can use in sequencers but differs by being integrated into the sampler and this provides some benefits. Unlike MIDI effects, scripts have access to some internal Kontakt features which in addition lets them:

    change volume/panning/tuning of a specific note
    fade in/out a specific note
    specify which groups should be used for playing back a specific note
    start playback of a note at any sample offset (only in Sampler mode)
    control a wide range of Kontakt parameters (maybe in the Kontakt 2.1 update)

Scripts are written in the Kontakt Script Language and entered as plain text. A script basically consists of a number text lines with instructions to the Kontakt Script Processor (KSP). The KSP interpretes your script and executes your instructions when it receives notes or MIDI CC."

From 
http://doc.linuxsampler.org/Instrument_Scripts/NKSP_Language/ 

"At a Glance

NKSP stands for "is Not KSP", which denotes its distinction to an existing proprietary language called KSP. NKSP is a script language specifically designed to write real-time capable software extensions to LinuxSampler's sampler engines that can be bundled individually with sounds by sound designers themselves. Instead of defining a completely new script language, NKSP is leaned on that mentioned properiatary script language. The biggest advantage is that sound designers and musicians can leverage the huge amount of existing KSP scripts which are already available for various purposes on the Internet, instead of being forced to write all scripts from scratch in a completely different language.

That also means however that there are some differences between those two languages. Some extensions have been added to the NKSP core language to make it a bit more convenient and less error prone to write scripts, and various new functions had to be added due to the large difference of the sampler engines and their underlying sampler format. Efforts have been made though to make NKSP as much compatible to KSP as possible. The NKSP documentation will emphasize individual differences in the two languages and function implementations wherever they may occur, to give you immediate hints where you need to take care of regarding compatibility issues when writing scripts that should be spawned on both platforms. "
